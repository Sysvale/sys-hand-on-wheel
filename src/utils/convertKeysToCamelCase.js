import camelCase from 'lodash.camelcase';
import isObject from 'lodash.isobject';

const isArray = (arg) => Array.isArray(arg);

const convertKeysToCamelCase = (data) => {
	if (isArray(data)) {
		return data.map((element) => {
			if ((isObject(element) || isArray(element)) && !(element instanceof File)) {
				return convertKeysToCamelCase(element);
			}
			return element;
		});
	}
	const newData = {};
	Object.keys(data).forEach((key) => {
		if ((isObject(data[key]) || isArray(data[key])) && !(data[key] instanceof File)) {
			newData[camelCase(key)] = convertKeysToCamelCase(data[key]);
		} else {
			newData[camelCase(key)] = data[key];
		}
	});

	return newData;
};

export default convertKeysToCamelCase;
