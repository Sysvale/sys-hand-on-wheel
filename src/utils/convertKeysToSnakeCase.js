import isObject from 'lodash.isobject';
import snakeCase from 'lodash.snakecase';

const isArray = (arg) => Array.isArray(arg);

const convertKeysToSnakeCase = (data) => {
	if (isArray(data)) {
		return data.map((element) => {
			if ((isObject(element) || isArray(element)) && !(element instanceof File)) {
				return convertKeysToSnakeCase(element);
			}
			return element;
		});
	}
	const newData = {};
	Object.keys(data).forEach((key) => {
		if ((isObject(data[key]) || isArray(data[key])) && !(data[key] instanceof File)) {
			newData[snakeCase(key)] = convertKeysToSnakeCase(data[key]);
		} else {
			newData[snakeCase(key)] = data[key];
		}
	});

	return newData;
};

export default convertKeysToSnakeCase;
