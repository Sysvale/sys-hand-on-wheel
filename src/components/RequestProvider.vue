<script>
import { h } from 'vue';
import swal from 'sweetalert2';
import get from 'lodash.get';
import convertKeysToCamelCase from '../utils/convertKeysToCamelCase';
import convertKeysToSnakeCase from '../utils/convertKeysToSnakeCase';
import generateKey from '../utils/generateKey';
import getFirstErrorMessage from '../utils/getFirstErrorMessage';

const SUCCESS_SWAL_DEFAULT_CONFIG = {
	title: 'Informações salvas com sucesso!',
	icon: 'success',
	text: '',
	showCloseButton: true,
	confirmButtonText: 'Ok',
};

const ERROR_SWAL_DEFAULT_CONFIG = {
	icon: 'error',
	title: 'Erro',
	text: 'mensagem de error',
	showCancelButton: true,
	showConfirmButton: false,
	cancelButtonText: 'Fechar',
};

export default {
	inject: ['$_requestObserver'],
	emits: ['success', 'error', 'success-feedback-ok', 'error-feedback-ok', 'success-feedback-cancel', 'error-feedback-cancel'],
	props: {
		vid: {
			type: String,
			default: () => generateKey(),
		},
		tag: {
			type: String,
			default: 'div',
		},
		service: {
			type: Function,
			required: true,
		},
		payload: {
			type: Object,
			default: () => ({}),
		},
		payloadResolver: {
			type: Function,
			default: (data) => convertKeysToSnakeCase(data),
		},
		dataResolver: {
			type: Function,
			default: (data) => convertKeysToCamelCase(data),
		},
		successSwalConfig: {
			type: Object,
			default: () => SUCCESS_SWAL_DEFAULT_CONFIG,
		},
		errorSwalConfig: {
			type: Object,
			default: () => ERROR_SWAL_DEFAULT_CONFIG,
		},
		errorFeedbackResolver: {
			type: Function,
			default: null,
		},
		successFeedbackResolver: {
			type: Function,
			default: null,
		},
		showSuccessFeedback: {
			type: Boolean,
			default: false,
		},
		hideErrorFeedback: {
			type: Boolean,
			default: false,
		},
		immediate: {
			type: Boolean,
			default: false,
		},
		forceResetError: {
			type: Boolean,
			default: false,
		},
		initialData: {
			default: null,
		},
	},

	data() {
		return {
			loading: false,
			failed: false,
			succeeded: false,
			error: null,
			data: this.initialData,
		};
	},

	computed: {
		requestState() {
			return {
				loading: this.loading,
				failed: this.failed,
				succeeded: this.succeeded,
				error: this.error,
				data: this.data,
			};
		}
	},

	watch: {
		forceResetError(newValue) {
			if (newValue) {
				this.error = null;
			}
		},
	},

	mounted() {
		if (this.immediate) {
			this.action();
		}
	},

	created() {
		if(this.$_requestObserver) {
			this.$_requestObserver.subscribe(this);
		}
	},

	beforeDestroy() {
		if(this.$_requestObserver) {
			this.$_requestObserver.unsubscribe(this);
		}
	},

	methods: {
		action(payloadFromArgs) {
			this.startRequest();
			const payload = payloadFromArgs || this.payload;
			this.service(this.payloadResolver(payload))
				.then(
					({ data }) => {
						this.data = this.dataResolver(data);
						this.succeeded = true;
						this.$emit('success', this.data);

						if (this.showSuccessFeedback) {
							if (this.successFeedbackResolver) {
								this.successFeedbackResolver({ vm: this, data: this.data });
								return;
							}
							swal.fire({
								...SUCCESS_SWAL_DEFAULT_CONFIG,
								...this.successSwalConfig,
							}).then(() => {
								this.$emit('success-feedback-ok', this.data);
							});
						}
					},
				).catch(
					(error) => {
						this.failed = true;
						this.error = error;
						console.error(error);
						this.$emit('error', error);

						if (!this.hideErrorFeedback) {
							const errorMessage = getFirstErrorMessage(
								get(error, 'response.data', null),
								'Um erro aconteceu... por favor, tente novamente. Se o erro persistir, contate o suporte.',
							);

							if (this.errorFeedbackResolver) {
								this.errorFeedbackResolver({ vm: this, error, errorMessage });
								return;
							}

							swal.fire({
								...ERROR_SWAL_DEFAULT_CONFIG,
								...this.errorSwalConfig,
								text: errorMessage,
							}).then((result) => {
								if (result.isDismissed) {
									this.$emit('error-feedback-cancel', error);
								}
								if (result.isConfirmed) {
									this.$emit('error-feedback-ok', error);
								}
							});
						}
					},
				).finally(() => {
					this.loading = false;
				});
		},

		loadingTextResolver(label, loadingLabel = 'Carregando...') {
			if (this.loading) {
				return loadingLabel;
			}
			return label;
		},

		startRequest() {
			this.loading = true;
			this.failed = false;
			this.succeeded = false;
			this.error = null;
		},
	},

	render() {
		const slotProvider = this.$slots;
		const slot = slotProvider.default({
			...this.requestState,
			action: this.action,
			loadingTextResolver: this.loadingTextResolver,
			errorMessage: getFirstErrorMessage(
				get(this.error, 'response.data', null),
				'Um erro aconteceu... por favor, tente novamente. Se o erro persistir, contate o suporte.',
			),
		});

		return h(this.tag, slot);
	},
};
</script>
