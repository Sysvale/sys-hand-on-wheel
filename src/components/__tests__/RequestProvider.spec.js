import { createLocalVue, mount } from '@vue/test-utils';
import flushPromises from 'flush-promises';
import convertKeysToCamelCase from '../../utils/convertKeysToCamelCase';
import convertKeysToSnakeCase from '../../utils/convertKeysToSnakeCase';
import RequestProvider from '../RequestProvider.vue';

const localVue = createLocalVue();

const responseDataMock = {
	message: 'success message',
};

const payloadMock = {
	id: '1',
};

const successfulServiceMock = jest.fn(() => Promise.resolve({
	data: responseDataMock,
}));
const failedServiceMock = jest.fn(() => Promise.reject(Error('test error')));
const payloadResolverMock = jest.fn((payload) => convertKeysToCamelCase(payload));
const dataResolverMock = jest.fn((data) => convertKeysToSnakeCase(data));
const successFeedbackResolverMock = jest.fn();
const errorFeedbackResolverMock = jest.fn();
const defaultComponent = '<div />';

test('Component renders correctly', async () => {
	const wrapper = mount(RequestProvider, {
		localVue,
		slots: {
			default: defaultComponent,
		},
		propsData: {
			service: successfulServiceMock,
		},
	});

	await flushPromises();

	expect(wrapper).toMatchSnapshot();
});

describe('Service', () => {
	test('is called on mount if immediate is true', async () => {
		expect.assertions(2);
		mount(RequestProvider, {
			localVue,
			slots: {
				default: defaultComponent,
			},
			propsData: {
				service: successfulServiceMock,
				immediate: false,
			},
		});

		await flushPromises();

		expect(successfulServiceMock).not.toHaveBeenCalled();

		mount(RequestProvider, {
			localVue,
			slots: {
				default: defaultComponent,
			},
			propsData: {
				service: successfulServiceMock,
				immediate: true,
			},
		});

		await flushPromises();

		expect(successfulServiceMock).toHaveBeenCalled();
	});

	test('is called correctly when action is triggered manually', async () => {
		expect.assertions(1);
		const wrapper = mount(RequestProvider, {
			localVue,
			slots: {
				default: defaultComponent,
			},
			propsData: {
				service: successfulServiceMock,
			},
		});

		await flushPromises();

		wrapper.vm.action(payloadMock);

		await flushPromises();

		expect(successfulServiceMock).toHaveBeenCalledWith(payloadMock);
	});
});

describe('Event', () => {
	test('is emitted correctly when a request has succeeded', async () => {
		const wrapper = mount(RequestProvider, {
			localVue,
			slots: {
				default: defaultComponent,
			},
			propsData: {
				service: successfulServiceMock,
				immediate: true,
			},
		});

		await flushPromises();

		expect(wrapper.emitted()).toHaveProperty('success');
	});

	test('is emitted correctly when a request has failed', async () => {
		const wrapper = mount(RequestProvider, {
			localVue,
			slots: {
				default: defaultComponent,
			},
			propsData: {
				service: failedServiceMock,
				immediate: true,
			},
		});

		await flushPromises();

		expect(wrapper.emitted()).toHaveProperty('error');
	});
});

describe('Resolvers', () => {
	test('have their results applied correctly to payload and data', async () => {
		const wrapper = mount(RequestProvider, {
			localVue,
			slots: {
				default: defaultComponent,
			},
			propsData: {
				service: successfulServiceMock,
				payloadResolver: payloadResolverMock,
				dataResolver: dataResolverMock,
				showSuccessFeedback: false,
				payload: payloadMock,
				immediate: true,
			},
		});

		expect(payloadResolverMock).toHaveBeenCalledWith(payloadMock);

		await flushPromises();

		expect(wrapper.vm.payload).toStrictEqual(convertKeysToCamelCase(payloadMock));

		expect(dataResolverMock).toHaveBeenCalledWith(responseDataMock);

		await flushPromises();

		expect(wrapper.vm.data).toStrictEqual(convertKeysToSnakeCase(responseDataMock));
	});

	test('are called correctly when a request has succeeded and showSuccessFeedback is false', async () => {
		expect.assertions(4);
		const wrapper = mount(RequestProvider, {
			localVue,
			slots: {
				default: defaultComponent,
			},
			propsData: {
				service: successfulServiceMock,
				payloadResolver: payloadResolverMock,
				dataResolver: dataResolverMock,
				successFeedbackResolver: successFeedbackResolverMock,
				showSuccessFeedback: false,
				payload: payloadMock,
				immediate: true,
			},
		});

		expect(payloadResolverMock).toHaveBeenCalledWith(payloadMock);

		await flushPromises();

		expect(dataResolverMock).toHaveBeenCalledWith(responseDataMock);
		expect(wrapper.emitted()).toHaveProperty('success');
		expect(successFeedbackResolverMock).not.toHaveBeenCalled();
	});

	test('are called correctly when a request has succeeded and showSuccessFeedback is true', async () => {
		expect.assertions(4);
		const wrapper = mount(RequestProvider, {
			localVue,
			slots: {
				default: defaultComponent,
			},
			propsData: {
				service: successfulServiceMock,
				payloadResolver: payloadResolverMock,
				dataResolver: dataResolverMock,
				successFeedbackResolver: successFeedbackResolverMock,
				showSuccessFeedback: true,
				payload: payloadMock,
				immediate: true,
			},
		});

		expect(payloadResolverMock).toHaveBeenCalledWith(payloadMock);

		await flushPromises();

		expect(dataResolverMock).toHaveBeenCalledWith(responseDataMock);
		expect(wrapper.emitted()).toHaveProperty('success');
		expect(successFeedbackResolverMock).toHaveBeenCalled();
	});

	test('are called correctly when a request has failed and hideErrorFeedback is true', async () => {
		expect.assertions(4);
		const wrapper = mount(RequestProvider, {
			localVue,
			slots: {
				default: defaultComponent,
			},
			propsData: {
				service: failedServiceMock,
				payloadResolver: payloadResolverMock,
				dataResolver: dataResolverMock,
				errorFeedbackResolver: errorFeedbackResolverMock,
				hideErrorFeedback: true,
				payload: payloadMock,
				immediate: true,
			},
		});

		expect(payloadResolverMock).toHaveBeenCalledWith(payloadMock);

		await flushPromises();

		expect(dataResolverMock).toHaveBeenCalledWith(responseDataMock);
		expect(wrapper.emitted()).toHaveProperty('error');
		expect(errorFeedbackResolverMock).not.toHaveBeenCalled();
	});

	test('are called correctly when a request has failed and hideErrorFeedback is false', async () => {
		expect.assertions(4);
		const wrapper = mount(RequestProvider, {
			localVue,
			slots: {
				default: defaultComponent,
			},
			propsData: {
				service: failedServiceMock,
				payloadResolver: payloadResolverMock,
				dataResolver: dataResolverMock,
				errorFeedbackResolver: errorFeedbackResolverMock,
				hideErrorFeedback: false,
				payload: payloadMock,
				immediate: true,
			},
		});

		expect(payloadResolverMock).toHaveBeenCalledWith(payloadMock);

		await flushPromises();

		expect(dataResolverMock).toHaveBeenCalledWith(responseDataMock);
		expect(wrapper.emitted()).toHaveProperty('error');
		expect(errorFeedbackResolverMock).toHaveBeenCalled();
	});
});

describe('Error state', () => {
	test('is reset after forceResetError is set to true', async () => {
		const wrapper = mount(RequestProvider, {
			localVue,
			slots: {
				default: defaultComponent,
			},
			propsData: {
				service: failedServiceMock,
				payload: payloadMock,
				immediate: true,
			},
		});

		await flushPromises();

		expect(wrapper.emitted()).toHaveProperty('error');
		expect(wrapper.vm.error).toBeTruthy();

		wrapper.setProps({
			forceResetError: true,
		});

		await flushPromises();

		expect(wrapper.vm.error).toBeFalsy();
	});
});

describe('Label', () => {
	test('changes correctly during request loading', async () => {
		expect.assertions(2);

		const wrapper = mount(RequestProvider, {
			localVue,
			slots: {
				default: defaultComponent,
			},
			propsData: {
				service: failedServiceMock,
				payload: payloadMock,
				immediate: true,
			},
		});

		expect(wrapper.vm.labelHelper('test label')).toBe('Carregando...');

		await flushPromises();

		expect(wrapper.vm.labelHelper('test label')).toBe('test label');
	});
});
