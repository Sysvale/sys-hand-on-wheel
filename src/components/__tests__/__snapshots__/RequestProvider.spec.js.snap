// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component renders correctly 1`] = `
<div>
  <div></div>
</div>
`;
